utils/intersection.go

```diff
@@ -0,0 +1,123 @@
+package utils
+
+// This file implements planar intersections (i.e. Flat Earth™)
+//
+// For spherical intersections, use LatLonSpherical.Intersection.
+// For rhumb line intersections, convert to Mercator projection using LatLon.Mercator*() and then use these planar functions.
+// TODO: combine all of the above provide functions that take an EarthModel argument
+
+import (
+	"github.com/starboard-nz/orb"
+	geod "github.com/starboard-nz/go-geodesy"
+)
+
+// LineStringIntersections finds the intersections of 2 LineStrings (if exists).
+func LineStringIntersections(l1, l2 orb.LineString) []orb.Point {
+	if len(l1) < 2 || len(l2) < 2 {
+		return nil
+	}
+
+	var intersections []orb.Point
+
+	for i := 1; i < len(l1); i++ {
+		for j := 1; j < len(l2); j++ {
+			is := SegmentIntersection(l1[i-1], l1[i], l2[j-1], l2[j])
+			if is != nil {
+				intersections = append(intersections, *is)
+			}
+		}
+	}
+
+	return intersections
+}
+
+// LineStringsIntersect returns true if the 2 LineStrings intersect.
+func LineStringsIntersect(l1, l2 orb.LineString) bool {
+	if len(l1) < 2 || len(l2) < 2 {
+		return false
+	}
+
+	for i := 1; i < len(l1); i++ {
+		for j := 1; j < len(l2); j++ {
+			if SegmentsIntersect(l1[i-1], l1[i], l2[j-1], l2[j]) {
+				return true
+			}
+		}
+	}
+
+	return false
+}
+
+// SegmentIntersection returns the intersections of 2 segments (p1, p2) and (q1, q2) (if exists).
+func SegmentIntersection(p1, p2, q1, q2 orb.Point) *orb.Point {
+	var p *orb.Point
+	_ = segmentIntersection(p1, p2, q1, q2, &p)
+
+	return p
+}
+
+// SegmentsIntersect returns true if segments (p1, p2) and (q1, q2) intersect.
+func SegmentsIntersect(p1, p2, q1, q2 orb.Point) bool {
+	return segmentIntersection(p1, p2, q1, q2, nil)
+}
+
+func segmentIntersection(p1, p2, q1, q2 orb.Point, is **orb.Point) bool {
```

Comment:
`is **orb.Point` I've never seen this before. How is this useful?

Reply:
allows returning a value of `nil` in `*is`, if it's not `nil`